### PR TITLE
Dolfyn updates + turbulence analysis

### PR DESCRIPTION
**Pull request by jmcvey3**
A slew of bugfixes and inconsistency cleanup, as well as including ADCP turbulence functionality for #227 

Changelog: 
  - Bugfixes
	  - Added check to ensure `n_bin` is shorter than the total data length when calling `dolfyn.TimeBinner.reshape`
	  - Added checks to ensure `n_fft` and `n_fft_coh` can't be greater than `n_bin`
	  - Fixed bug where `dolfyn.adp.nan_beyond_surface` overtrimmed TRDI instrument data
	  - Fixed bug where `dolfyn.ADVBinner.cross_spectral_density` would fail if `n_fft` != `n_fft_coh`
	  - Improved robustness of Nortek Signature reader for deployments not measuring water velocity
	  - log file extension changed from ".log" to ".dolfyn.log"
	  
  - API/Useability
	  - Calculation of depth from pressure sensor updated to use linear approximation of the equation of state, rather than EOS-80
	  - Added warnings for ADV motion and turbulence functions
	  - Updated `dataset.velds.U_dir` shortcut to automatically convert "degrees CCW from X/East/streamwise" to "degrees CW from X/North/streamwise"
	  - `dolfyn.ADVBinner.cross_spectral_density` now returns frequency coordinate `coh_freq` instead of `freq`
	  - Added ADCP turbulence functions
	  - Added function to calculate Doppler noise to ADV turbulence functions
	  - Added "beam_angle" attribute to Nortek Signature datasets
	  - Saved full Nortek Signature "config" dictionary as json string in attributes
	  - Added warning if "rotate_vars" attribute not found
	  - Standardized variable metadata based on CF conventions

### CODE REVIEW

**Comment by ssolson** [comment]
James pulling the latest develop branch should fix your test failures

**Comment by jmcvey3** [comment]
Currently working on adding turbulence functions to ADCP example notebook

**Comment by ssolson** [comment]
@jmcvey3 in #236 you said:
>  @ssolson I just remembered there are some changes in the dolfyn PR (namely to the function/shortcut dolfyn.velocity.U_dir) that will affect this one. I'll try to wrap the dolfyn PR soon here so we can merge these back to back.

So we need to complete this PR first then #236? I am only confused because it seems #236 has passed all tests.

**Comment by jmcvey3** [comment]
@ssolson Just pushed some updates to the jupyter notebooks to incorporate the ADCP turbulence analysis. 

> @jmcvey3 in #236 you said:
> 
> > @ssolson I just remembered there are some changes in the dolfyn PR (namely to the function/shortcut dolfyn.velocity.U_dir) that will affect this one. I'll try to wrap the dolfyn PR soon here so we can merge these back to back.
> 
> So we need to complete this PR first then #236? I am only confused because it seems #236 has passed all tests.

Just to have this on paper, it's a change relevant to the tidal performance jupyter notebook

**Comment by jmcvey3** [comment]
@ssolson Just pushed some updates to the jupyter notebooks to incorporate the ADCP turbulence analysis, and this PR should be good to go for final review

File: mhkit/dolfyn/adp/turbulence.py [hunk 1]
```diff
@@ -0,0 +1,886 @@
...
+            or 'centered_extended' (centered difference with first and last points
+             extended using a first difference).
+        """
+
+        VelBinner.__init__(self, n_bin, fs, n_fft, n_fft_coh, noise)
+        self.diff_style = diff_style
+        self.orientation = orientation
+
+    def _diff_func(self, vel, u):
+        if self.diff_style == 'first':
```
Comment (ssolson) [inline mhkit/dolfyn/adp/turbulence.py:72]: Is this docstring correct?
```
    """
    Applies the chosen style of numerical differentiation to velocity data.

    This method calculates the derivative of the velocity data 'vel' with respect to the 'range' 
    using the differentiation style specified in 'self.diff_style'. The styles can be 'first' 
    for first difference, 'centered' for centered difference, and 'centered_extended' for 
    centered difference with first and last points extended using a first difference.

    Parameters
    ----------
    vel : xarray.Dataset
        An xarray Dataset containing the velocity data. This should have a 'range' coordinate.

    u : str
        A string specifying the velocity component to differentiate. This should be a variable in 'vel'.

    Returns
    -------
    out : numpy.ndarray
        The calculated derivative of the velocity data.

    range_out : xarray.DataArray
        The 'range' coordinate corresponding to 'out'. Its length may be less than the length of 'vel.range' 
        depending on the differentiation style.
    """
    ```

File: mhkit/dolfyn/adp/turbulence.py [hunk 2]
```diff
@@ -0,0 +1,886 @@
...
+
+        Parameters
+        ----------
+        vel : xarray.DataArray
+          ADCP raw velocity
+        orientation : str, default=ADPBinner.orientation
+          Direction ADCP is facing ('up' or 'down')
+
+        Notes
+        -----
```
Comment (ssolson) [inline mhkit/dolfyn/adp/turbulence.py:94]: Can you add a Returns here?

File: mhkit/dolfyn/adp/turbulence.py [hunk 3]
```diff
@@ -0,0 +1,886 @@
...
+
+    def dvdz(self, vel):
+        """
+        The shear in the second velocity component.
+
+        Parameters
+        ----------
+        vel : xarray.DataArray
+          ADCP raw velocity
+
```
Comment (ssolson) [inline mhkit/dolfyn/adp/turbulence.py:122]: Can you add a Returns here?

File: mhkit/dolfyn/adp/turbulence.py [hunk 4]
```diff
@@ -0,0 +1,886 @@
...
+    def dwdz(self, vel):
+        """
+        The shear in the third velocity component.
+
+        Parameters
+        ----------
+        vel : xarray.DataArray
+          ADCP raw velocity
+
```
Comment (ssolson) [inline mhkit/dolfyn/adp/turbulence.py:146]: Can you add a Returns here?

File: mhkit/dolfyn/adp/turbulence.py [hunk 5]
```diff
@@ -0,0 +1,886 @@
...
+    def shear_squared(self, vel):
+        """
+        The horizontal shear squared.
+
+        Parameters
+        ----------
+        vel : xarray.DataArray
+          ADCP raw velocity
+
```
Comment (ssolson) [inline mhkit/dolfyn/adp/turbulence.py:170]: Can you add a Returns here?

File: mhkit/dolfyn/adp/turbulence.py [hunk 6]
```diff
@@ -0,0 +1,886 @@
...
+            noise_level.values.astype('float32'),
+            dims=['time'],
+            attrs={'units': 'm s-1',
+                   'long_name': 'Doppler Noise Level',
+                   'description': 'Doppler noise level calculated '
+                   'from PSD white noise'})
+
+    def _stress_func_warnings(self, ds, beam_angle, noise, tilt_thresh):
+        """List of error and warnings to run through for ADCP stress calculations.
+        """
```
Comment (ssolson) [inline mhkit/dolfyn/adp/turbulence.py:259]: Is this docstring correct?

```
    """
    Performs a series of checks and raises warnings for ADCP stress calculations.

    This method checks several conditions relevant for ADCP stress calculations and raises 
    warnings if these conditions are not met. It checks if the beam angle is defined, 
    if the instrument's coordinate system is aligned with the principal flow directions, 
    if the tilt is above a threshold, if the noise level is specified, and if the data 
    set is in the 'beam' coordinate system.

    Parameters
    ----------
    ds : xarray.Dataset
        An xarray Dataset containing the ADCP data.

    beam_angle : float
        The beam angle of the ADCP in degrees.

    noise : float
        The noise level of the ADCP.

    tilt_thresh : float
        The tilt threshold in degrees. If the tilt exceeds this value, a warning is raised.

    Returns
    -------
    b_angle : float
        The actual beam angle used for the calculations. If 'beam_angle' was None, it tries to find it in 'ds'.

    noise : float
        The actual noise level used for the calculations. If 'noise' was None, it is set to 0.
    """
    ```

File: mhkit/dolfyn/adp/turbulence.py [hunk 7]
```diff
@@ -0,0 +1,886 @@
...
+        Get the beam order for the beam-stress rotation algorithm
+
+        Note: Stacey defines the beams for down-looking Workhorse ADCPs.
+              According to the workhorse coordinate transformation
+              documentation, the instrument's:
+                               x-axis points from beam 1 to 2, and
+                               y-axis points from beam 4 to 3.
+        Nortek Signature x-axis points from beam 3 to 1
+                         y-axis points from beam 2 to 4
+        """
```
Comment (ssolson) [inline mhkit/dolfyn/adp/turbulence.py:304]: Consider working your current docstring into this outline:
```
    """
    Determines the beam order for the beam-stress rotation algorithm based on the instrument orientation.

    This method checks the orientation of the instrument and adjusts the beam order accordingly. 
    The beam order is defined differently for different types of instruments (e.g., TRDI or Nortek Signature) 
    and orientations (up or down).

    Parameters
    ----------
    ds : xarray.Dataset
        An xarray Dataset containing the ADCP data.

    orientation : str
        The orientation of the instrument, either 'up' or 'down'. 
        If None, the orientation will be retrieved from the dataset or the instance's default orientation.

    beam5 : bool, default=False
        A flag indicating whether a fifth beam is present. If True, the number 4 will be appended to the beam order.

    Returns
    -------
    beams : list of int
        A list of integers representing the order of the beams.

    phi2 : float, optional
        The mean of the roll values in radians. Only returned if 'beam5' is True.

    phi3 : float, optional
        The mean of the pitch values in radians, negated for Nortek instruments. 
        Only returned if 'beam5' is True.
    """
    ```

File: mhkit/dolfyn/adp/turbulence.py [hunk 8]
```diff
@@ -0,0 +1,886 @@
...
+
+        if beam5:
+            beams.append(4)
+            return beams, phi2, phi3
+        else:
+            return beams
+
+    def _beam_variance(self, ds, time, noise, beam_order, n_beams):
+        """Calculate along-beam velocity variance and subtract noise
+        """
```
Comment (ssolson) [inline mhkit/dolfyn/adp/turbulence.py:341]: Is this docsting correct?

```
    """
    Determines the beam order for the beam-stress rotation algorithm based on the instrument orientation.

    This method checks the orientation of the instrument and adjusts the beam order accordingly. 
    The beam order is defined differently for different types of instruments (e.g., TRDI or Nortek Signature) 
    and orientations (up or down).

    Parameters
    ----------
    ds : xarray.Dataset
        An xarray Dataset containing the ADCP data.

    orientation : str
        The orientation of the instrument, either 'up' or 'down'. 
        If None, the orientation will be retrieved from the dataset or the instance's default orientation.

    beam5 : bool, default=False
        A flag indicating whether a fifth beam is present. If True, the number 4 will be appended to the beam order.

    Returns
    -------
    beams : list of int
        A list of integers representing the order of the beams.

    phi2 : float, optional
        The mean of the roll values in radians. Only returned if 'beam5' is True.

    phi3 : float, optional
        The mean of the pitch values in radians, negated for Nortek instruments. 
        Only returned if 'beam5' is True.
    """
    ```

**Review by ssolson — COMMENTED** [review]
James you did  a lot of great work here. I have a number of minor suggestions primarily around docstrings. Outside of the new turbulence model this seems to be a great deal of clean up and improvements top the current dolfyn module.

I  still need to review the notebooks next week.

File: mhkit/dolfyn/rotate/base.py [hunk 9]
```diff
@@ -298,6 +317,37 @@ def quaternion2orient(quaternions):
...
+    if 'xarray' in type(pitch).__module__:
+        pitch = pitch.values
+    if 'xarray' in type(roll).__module__:
+        roll = roll.values
+
+    tilt = np.arctan(
+        np.sqrt(np.tan(np.deg2rad(roll)) ** 2 + np.tan(np.deg2rad(pitch)) ** 2)
+    )
+
+    return np.rad2deg(tilt)
```
[the remark this answers is not in the document]
Comment (jmcvey3) [reply mhkit/dolfyn/rotate/base.py:353]: Yes, tilt for ADCPs is the vertical inclination and ranges from +/- 90 degrees for pitch and roll

File: mhkit/dolfyn/adp/turbulence.py [hunk 10]
```diff
@@ -0,0 +1,886 @@
+import numpy as np
+import xarray as xr
+import warnings
+
+from ..velocity import VelBinner
+from ..rotate.base import calc_tilt
+
+
+def _diffz_first(dat, z):
+    """Newton's Method first difference.
```
[the remark this answers is not in the document]
Comment (jmcvey3) [reply mhkit/dolfyn/adp/turbulence.py:10]: As much as I enjoy the chatbot generated docstrings, I don't see the need to over-explain the Newton's methods defined here.

**Comment by ssolson** [comment]
@jmcvey3 please ping me when you are ready for me to take a look. On this PR I have https://github.com/jmcvey3/MHKiT-Python/pull/4 open agains your branch for my suggestions for the ipython notebook.

**Comment by jmcvey3** [comment]
@ssolson Ready to go! I cleaned up the rest of the docstrings (I think I got everything) and added more checks for function inputs.

**Comment by ssolson** [comment]
James thank you making all those changes. 

Currently the CI/CD has passed all but the hindcast which this PR does not modify. As such I am going to proceed with the merge.

**Review by ssolson — APPROVED** [review]

**Pull request merged**